### PR TITLE
[All devs approval required] Developer workflow document

### PR DESCRIPTION
**Pull request by arjunu**
The goal of this is to: 
- make explicit all conventions we follow - branch name format, commit messages format etc.
- acceptability criteria for a PR - does it need unit tests or e2e tests or both, when to add stories

### CODE REVIEW

File: docs/developer-workflow.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+// TODO define acceptability criteria <br/>
+// what needs unit tests <br/>
+// what needs e2e tests <br/>
+// what needs stories <br/>
+// adding missing types
+
+##### Documentation
+
+// TODO
```
Comment (desfero) [inline docs/developer-workflow.md]:
```suggestion
As for now, we don't have any strict guidelines for documentation. 
It's up for a developer to decide when to document a given part of the system of but we are trying to follow points mentioned below:
- If the function is pretty complex first of all try to split it by smaller parts more focus parts. i=Ff that's not possible add `tsdoc` comment with a short (but comprehensive) explanation.
- If you need to do something unusual (for .e.g. force cast type) add a short comment why above that line.
```

File: docs/developer-workflow.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+### 3. Create a PR
+
+For WIP PRs, you can use a draft PR or mark it as "[WIP]". Using Zenhub, link the PR to the corresponding issue (if exists).
+
+### 4. Dev Review
+
+###### Board status: Review
+
+Once done, request for
+a dev review in the front end channel on Slack.
```
Comment (desfero) [inline docs/developer-workflow.md]:
```suggestion
a dev review in the front end channel on Slack.
We require at least one approval from the dev to push PR to the **QA Review**.
 
In case the PR is related to the dependencies update and it introduces new features we need to be aware of (for .e.g. react update which introduces hooks) then all devs approval is required before merging and PR should contain external links to changelogs. Often PR is marked as `[All devs approval required]` in the title. 
```

File: docs/developer-workflow.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+###### Board status: QA
+
+This step can be skipped if it's a purely technical task and does not change the behavior of the application e.g. upgrading dependencies, refactor etc. Mark such PRs with a `Tech` label.
+
+### 6. Merge
+
+###### Board status: Staging
+
+You can squash merge if you don't want to keep the commits or fast forward if otherwise.
```
Comment (DmitriyNoa) [inline docs/developer-workflow.md]: Looks like squash and merge is not optional when merging to master. Let's make it required

Comment (desfero) [reply docs/developer-workflow.md]: Right, the flow is the following:
If it's a feature PR we do almost 99% squash to `master` branch
If it's a release pr (from `master` or any other branch to `production` then we always use `merge`)

File: docs/developer-workflow.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Developer Workflow [internal]
+
+### 1. Create branch for the feature/issue/task
+
+###### Board status: In Progress
+
+Branch name convention is `{username}/{issueid}-{short-description}`
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Add information if its a hotfix or not

File: docs/developer-workflow.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+### 2. Code
+
+##### Tests
+
+// TODO define acceptability criteria <br/>
+// what needs unit tests <br/>
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: If its a util function then we add unit tests. 
- From now on
any new module to the sagas

File: docs/developer-workflow.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+// what needs e2e tests <br/>
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Anything that has a user or a flow should be added as an E2E test. 
This includes the happy path, and expected error handeling. and if there are any edge cases. 

Write more about when should the test be `io` and when not

File: docs/developer-workflow.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+// what needs stories <br/>
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Any UI component needs a story 

Container -> should have a story
-------------
ConnectedComponent

File: docs/developer-workflow.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+// adding missing types
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: For sagas any saga you edit and has any you are responsible for typing it. 

For `.unsafe.tx` they include but might have a complicated. 

Add link to saga migration plan

File: docs/developer-workflow.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+// when to create selectors
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: To discuss in the module workflow

File: docs/developer-workflow.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+// where to create css files (container or child component)
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Don't pass scss and style from Pages 

When we have a full design system we shouldn't style any reusable components like buttons etc..

File: docs/developer-workflow.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+##### Documentation
+
+As for now, we don't have any strict guidelines for documentation. 
+It's up for a developer to decide when to document a given part of the system of but we are trying to follow points mentioned below:
+- If the function is pretty complex first of all try to split it by smaller parts more focus parts. If that's not possible add `tsdoc` comment with a short (but comprehensive) explanation.
+- If you need to do something unusual (for .e.g. force cast type) add a short explanation above that line.
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Add a `comment`

File: docs/developer-workflow.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+// TODO define acceptability criteria <br/>
+// what needs unit tests <br/>
+// what needs e2e tests <br/>
+// what needs stories <br/>
+// adding missing types
+// when to create selectors
+// where to create css files (container or child component)
+
+##### Documentation
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Link the readme md here

File: docs/developer-workflow.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+As for now, we don't have any strict guidelines for documentation. 
+It's up for a developer to decide when to document a given part of the system of but we are trying to follow points mentioned below:
+- If the function is pretty complex first of all try to split it by smaller parts more focus parts. If that's not possible add `tsdoc` comment with a short (but comprehensive) explanation.
+- If you need to do something unusual (for .e.g. force cast type) add a short explanation above that line.
+
+### 3. Create a PR
+
+For WIP PRs, you can use a draft PR or mark it as "[WIP]". Using Zenhub, link the PR to the corresponding issue (if exists).
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: If you have any special issues or remarks

File: docs/developer-workflow.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+###### Board status: In Progress
+
+Branch name convention is `{username}/{issueid}-{short-description}`
+
+### 2. Code
+
+##### Tests
+
+// TODO define acceptability criteria <br/>
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: If you import some external code from stackoverflow or 
please add a comment and link to it

File: docs/developer-workflow.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+// where to create css files (container or child component)
+
+##### Documentation
+
+As for now, we don't have any strict guidelines for documentation. 
+It's up for a developer to decide when to document a given part of the system of but we are trying to follow points mentioned below:
+- If the function is pretty complex first of all try to split it by smaller parts more focus parts. If that's not possible add `tsdoc` comment with a short (but comprehensive) explanation.
+- If you need to do something unusual (for .e.g. force cast type) add a short explanation above that line.
+
+### 3. Create a PR
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Tech: this means that QA doesn't need to review.
QA: is done by the code reviewer 
E2E: for pr that touch only e2e tests

File: docs/developer-workflow.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+For WIP PRs, you can use a draft PR or mark it as "[WIP]". Using Zenhub, link the PR to the corresponding issue (if exists).
+
+### 4. Dev Review
+
+###### Board status: Review
+
+Once done, request for dev review in the front end channel on Slack. We require at least one approval from the dev to push PR to the **QA Review**.
+
+In case a PR involves dependency updates or if it introduces changes other devs need to be aware of (for .e.g. React update which introduces hooks) then request approval from all devs. The PR should contain external links to changelogs if there are any. Such kind of a PR must be marked with `[All devs approval required]` in the title.
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: The idea behind all devs approving is coming from a more informational point. 

The first guy reviewing should review also the code. Anyone else has the choice

File: docs/developer-workflow.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+For this:
+- the branch has to be brought up to date with master. Rebase it on master using fast-forward mode to avoid merge commits.
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Before you set the PR for review make sure you are merged with the latest master. 
Not necessary fast-forward because we will squash in the end

File: docs/developer-workflow.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+You can squash merge if you don't want to keep the commits or fast forward if otherwise.
+Merge commit format convention:
+
+`(#${issue-number}) {message} (#{PR-number})`
+
+Once reviewed, merge the PR. If it needs more work, move it back to In Progress on the board.
+
+### 7. Closed
+
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Add Documentation on how to do the release.

File: docs/developer-workflow.md [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+This step can be skipped if it's a purely technical task and does not change the behavior of the application e.g. upgrading dependencies, refactor etc. Mark such PRs with a `Tech` label.
+
+### 6. Merge
+
+###### Board status: Staging
+
+You can squash merge if you don't want to keep the commits or fast forward if otherwise.
+Merge commit format convention:
+
+`(#${issue-number}) {message} (#{PR-number})`
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Remove ` (#{PR-number})`  as its added automatically

File: docs/developer-workflow.md [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+###### Board status: Review
+
+Once done, request for dev review in the front end channel on Slack. We require at least one approval from the dev to push PR to the **QA Review**.
+
+In case a PR involves dependency updates or if it introduces changes other devs need to be aware of (for .e.g. React update which introduces hooks) then request approval from all devs. The PR should contain external links to changelogs if there are any. Such kind of a PR must be marked with `[All devs approval required]` in the title.
+
+For this:
+- the branch has to be brought up to date with master. Rebase it on master using fast-forward mode to avoid merge commits.
+- All checks must have passed
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Write about the process when the pipeline is blocked tests failing outside of the context of PRs

File: docs/developer-workflow.md [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+Keep an eye out for change in
+- Build size
+- Type coverage
+
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: Visual regression 
Full visual regression

File: docs/developer-workflow.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+### 5. QA Review
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: At somepoint the QA needs to approve and we merge or ask for changes and ticket goes back to `in progress`

File: docs/developer-workflow.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+###### Board status: Staging
+
+You can squash merge if you don't want to keep the commits or fast forward if otherwise.
+Merge commit format convention:
+
+`(#${issue-number}) {message} (#{PR-number})`
+
+Once reviewed, merge the PR. If it needs more work, move it back to In Progress on the board.
+
+### 7. Closed
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: The ticket is only closed once its released on production
And will be closed by QA 

Developers can't touch the ticket after it goes from `review` to ` QA` 
Except Tech Tickets then push directly to `stage`

File: docs/developer-workflow.md (hunk 13 above)
Comment (DmitriyNoa) [inline docs/developer-workflow.md]: Make sure to run yarn lint:fix before creating a pr

File: docs/developer-workflow.md [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+### 5. QA Review
+
+###### Board status: QA
+
+This step can be skipped if it's a purely technical task and does not change the behavior of the application e.g. upgrading dependencies, refactor etc. Mark such PRs with a `Tech` label.
+
+### 6. Merge
+
+###### Board status: Staging
```
Comment (DmitriyNoa) [inline docs/developer-workflow.md]: If the branch has conflicts with the master in packages/web/health-check-sha.json always take the latest SHA from the branch, assuming the yarn lint:fix has been executed

File: docs/developer-workflow.md [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+- See [React guidelines](https://github.com/Neufund/platform-frontend/blob/master/docs/react-guidelines.md)
+- See [Working with Intl](https://github.com/Neufund/platform-frontend/blob/master/docs/working-with-Intl.md)
+``
+#### Acceptability criteria for a PR:<br/>
+
+- UI components must be accompanied by stories (Atoms/Molecules/Organisms but not Pages/Connected Components)
+
+##### Tests
+- for utils, add unit tests
+- anything that modifies a user flow must be accompanies by E2E tests (he happy path, error handling path and any edge cases)
```
Comment (pavelneu) [inline docs/developer-workflow.md]:
```suggestion
- anything that modifies a user flow must be accompanies by E2E tests (the happy path, error handling path and any edge cases)
```

File: docs/developer-workflow.md [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+- See [React guidelines](https://github.com/Neufund/platform-frontend/blob/master/docs/react-guidelines.md)
+- See [Working with Intl](https://github.com/Neufund/platform-frontend/blob/master/docs/working-with-Intl.md)
+``
+#### Acceptability criteria for a PR:<br/>
+
+- UI components must be accompanied by stories (Atoms/Molecules/Organisms but not Pages/Connected Components)
+
+##### Tests
+- for utils, add unit tests
+- anything that modifies a user flow must be accompanied by E2E tests (the happy path, error handling path and any edge cases)
```
Comment (pavelneu) [inline docs/developer-workflow.md]:
```suggestion
- anything that modifies a user flow must be accompanied by E2E tests (the happy path, error handling path and any edge cases)
- for reducing flakiness: The tests should pass at least 3 (three) times in a row.
```

File: docs/developer-workflow.md [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+
+##### Adding Types
+- type any sagas your PR touches. See [Saga Migration Plan](https://github.com/Neufund/platform-frontend/blob/master/docs/saga-migration-guidelines.md)
+- files named `.unsafe.tsx` may already contain types but they must be fully typed
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: if the file is `unsafe` it contains `any` type. 
we shouldn't use `any` at all

File: docs/developer-workflow.md [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+
+#### Documentation
+
+As for now, we don't have any strict guidelines for documentation.
+It's up for a developer to decide when to document a given part of the system of but we are trying to follow points mentioned below:
+- If the function is pretty complex first of all try to split it by smaller parts more focus parts. If that's not possible add `tsdoc` comment with a short (but comprehensive) explanation.
+- If you need to do something unusual (for .e.g. force cast type) add a short explanation above that line.
+- When adding a new package, add a README explaing what the package is about and other helpful information
+
```
Comment (Moejoe90) [inline docs/developer-workflow.md]: If we are gonna add new `modules` (saga+ reducer+ selector+ types+ spec) we need to have docoumentation on what this modules.
